### ISSUE TITLE
Issue for Self-signed certificates with files

### ISSUE THREAD

**Issue by preraksola** [Steinblock/redmine-gitlab-migrator#3]
Hey,

First of all, I modified the your code a bit in order to support `Self signed certificates` by default. So `no-verify` is by default set to `true` so there's no need to pass it during execution.
Also, I turned off the user ceck by commenting the line ` (check_users, 'Required users presence'),` from `commands.py`.

For migrating the issues, I executed the below command first:

```
migrate-rg issues --redmine-key xxxx --gitlab-key yyyy \
https://redmine.xx.net/projects/projectName https://gitlab.xx.com/zz/projectName --check
```

And the output is as below:
```
INFO: Project has no pre-existing issue...
INFO: Project has no pre-existing issue... OK
INFO: Getting redmine issues
INFO: Converting issues
INFO: Creating gitlab issues
WARNING: Redmine issue #783 is anonymous, gitlab issue is attributed to current admin

WARNING: Redmine user {'name': 'Admin', 'id': 1} is unknown, attribute note to current admin

INFO: Would create issue "-RM-783-MR-Test Issue" and 1 notes.
INFO: Would create issue "-RM-648-MR-Compatibility with X" and 0 notes.
INFO: Would create issue "-RM-636-MR-[Android] French Translation" and 0 notes.
INFO: Would create issue "-RM-629-MR-[Android] Déploiement sur le store" and 0 notes.
INFO: Would create issue "-RM-628-MR-[Android] Tests" and 0 notes.
INFO: Would create issue "-RM-627-MR-[Android] Modification du code" and 0 notes.
INFO: Would create issue "-RM-626-MR-[Android] Environnement de dev" and 0 notes.
```

After that I executed the below command to perform the migration:
```
migrate-rg issues --redmine-key xxxx --gitlab-key yyyy \
https://redmine.xx.net/projects/projectName https://gitlab.xx.com/zz/projectName
```

And the  output of it is as below:

```
INFO: Project has no pre-existing issue...
INFO: Project has no pre-existing issue... OK
INFO: Getting redmine issues
INFO: Converting issues
INFO: Creating gitlab issues
WARNING: Redmine issue #783 is anonymous, gitlab issue is attributed to current admin

WARNING: Redmine user {'name': 'Admin', 'id': 1} is unknown, attribute note to current admin

INFO: #17 -RM-783-MR-Test Issue
INFO: #18 -RM-648-MR-Compatibility with X
INFO: 	uploading ZeeOTP-Translation.xls (https://redmine.xx.net/attachments/download/239/abc-Translation.xls?key=xxxx / application/vnd.ms-excel)
INFO: create issue "-RM-636-MR-[Android] French Translation" failed
Traceback (most recent call last):
  File "/usr/lib/python3.5/urllib/request.py", line 1254, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "/usr/lib/python3.5/http/client.py", line 1106, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.5/http/client.py", line 1151, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.5/http/client.py", line 1102, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/usr/lib/python3.5/http/client.py", line 877, in send
    self.connect()
  File "/usr/lib/python3.5/http/client.py", line 1260, in connect
    server_hostname=server_hostname)
  File "/usr/lib/python3.5/ssl.py", line 377, in wrap_socket
    _context=self)
  File "/usr/lib/python3.5/ssl.py", line 752, in __init__
    self.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 988, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 633, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/migrate-rg", line 11, in <module>
    load_entry_point('redmine-gitlab-migrator==1.0.3', 'console_scripts', 'migrate-rg')()
  File "/usr/local/lib/python3.5/dist-packages/redmine_gitlab_migrator-1.0.3-py3.5.egg/redmine_gitlab_migrator/commands.py", line 308, in main
  File "/usr/local/lib/python3.5/dist-packages/redmine_gitlab_migrator-1.0.3-py3.5.egg/redmine_gitlab_migrator/commands.py", line 191, in perform_migrate_issues
  File "/usr/local/lib/python3.5/dist-packages/redmine_gitlab_migrator-1.0.3-py3.5.egg/redmine_gitlab_migrator/gitlab.py", line 135, in create_issue
  File "/usr/local/lib/python3.5/dist-packages/redmine_gitlab_migrator-1.0.3-py3.5.egg/redmine_gitlab_migrator/gitlab.py", line 103, in uploads_to_string
  File "/usr/lib/python3.5/urllib/request.py", line 163, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.5/urllib/request.py", line 466, in open
    response = self._open(req, data)
  File "/usr/lib/python3.5/urllib/request.py", line 484, in _open
    '_open', req)
  File "/usr/lib/python3.5/urllib/request.py", line 444, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 1297, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/usr/lib/python3.5/urllib/request.py", line 1256, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)>
```

As you can see, two issues are migrated successfully but for the third one which has an excel file, it gets the error.

Also, instead of self-signed certificate if the certificate is issued by a CA not in the trusted list or if the intermediate certificate is missing, can it lead to the above behavior?

Let me know if you need any more information from my side.

Prerak

**Comment by Steinblock** [comment]
@preraksola 

> First of all, I modified the your code a bit in order to support Self signed certificates by default. So no-verify is by default set to true so there's no need to pass it during execution.

This might be the problem. The argument is defined as:

```
i.add_argument(
    '--no-verify',
    required=False, action='store_false', default=True,
    help="disable SSL certificate verification")
```
The default is `True` and setting the argument sets the value to `False` (action=store_false)

I just pass the value (true or false) to the Redmine/GitLab api clients

```
redmine = RedmineClient(args.redmine_key, args.no_verify)
gitlab = GitlabClient(args.gitlab_key, args.no_verify)
```

which inherit APIClient the signture

```
class APIClient:
    def __init__(self, api_key, verify):
        self.api_key = api_key
        self.verify = verify
```

(Notice: it's `verify` not `no_verify` anymore and this is just passed to the internally used [requests api](http://docs.python-requests.org/en/master/api/)
> verify -- (optional) whether the SSL cert will be verified. A CA_BUNDLE path can also be provided. Defaults to True.

In other words: You have to change your code to `GitlabClient(args.gitlab_key, False)` to disable SSL certificate checking


I agree, this is a bit misleading, so I might change that for my pull request to:

```
 '--no-verify',
required=False, action='store_true', default=False,
help="disable SSL certificate verification")
...
    redmine = RedmineClient(args.redmine_key, not args.no_verify)
    gitlab = GitlabClient(args.gitlab_key, not args.no_verify)
```

**Comment by preraksola** [comment]
@Steinblock 
I restored the parameter settings as it is in your branch. And upon passing `--no-verify` parameter, I still experience the above behavior.
I also changed the values passed to api clients as below in `commands.py`, function `perform_migrate_issues`:
```
redmine = RedmineClient(args.redmine_key, False)
gitlab = GitlabClient(args.gitlab_key, False)
```

But the behavior is still the same as before.

**Comment by Steinblock** [comment]
Well, since attachments are uploaded via

```
upload = self.api.post(uploads_url, files=files)
```

which will set, `verify = False` this cannot be the problem. This might be caused by urlopen

```
files = [("file", (u['filename'], urlopen(u['content_url']), u['content_type']))]
```
but with python `3.4.3` on ubuntu 16.04 and self signed certificates for both redmine and gitlab, this did not happen to me.

according to this [stackoverflow post](http://stackoverflow.com/a/28052583/98491) this might be a [security fix](https://www.python.org/dev/peps/pep-0476/) applied in python 3.4.x

You could try the following code (gitlab.py ~100)

```
context = ssl._create_unverified_context()
files = [("file", (u['filename'], urlopen(u['content_url'], context=context), u['content_type']))]
```

Please note that there is another call to `urlopen` inside the catch block which also needs the context.

Please let me now if this resolves the issue.

**Comment by preraksola** [comment]
@Steinblock Yeah that did the trick. It is working fine now. 
Thanks :+1:
